File: frontend/src/api/wallet.ts
/**
 * Wallet API client — Digital Wallet & Subscription System (§6.60).
 */
import { api } from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface WalletData {
  id: number;
  package: string;
  package_credits: number;
  purchased_credits: number;
  total_credits: number;
  auto_refill_enabled: boolean;
  auto_refill_threshold_cents: number;
  auto_refill_amount_cents: number;
}

export interface PackageTier {
  id: number;
  name: string;
  monthly_credits: number;
  price_cents: number;
  is_active: boolean;
}

export interface CreditPackageItem {
  id: number;
  name: string;
  credits: number;
  price_cents: number;
}

export interface WalletTransactionItem {
  id: number;
  transaction_type: string;
  amount: number;
  balance_after: number;
  reference_id: string | null;
  payment_method: string | null;
  note: string | null;
  created_at: string;
}

export interface TransactionList {
  items: WalletTransactionItem[];
  total: number;
}

export interface CheckoutResponse {
  client_secret: string;
  publishable_key: string;
}

// ---------------------------------------------------------------------------
// API calls
// ---------------------------------------------------------------------------

export const walletApi = {
  /** Get current user's wallet balance and package info. */
  getWallet: () =>
    api.get<WalletData>('/api/wallet').then((r) => r.data),

  /** Get paginated transaction ledger. */
  getTransactions: (skip = 0, limit = 50) =>
    api
      .get<TransactionList>('/api/wallet/transactions', {
        params: { skip, limit },
      })
      .then((r) => r.data),

  /** List active package tiers. */
  getPackages: () =>
    api.get<PackageTier[]>('/api/wallet/packages').then((r) => r.data),

  /** Enroll in or change package tier. */
  enrollPackage: (packageName: string) =>
    api
      .post<WalletData>('/api/wallet/packages/enroll', {
        package_name: packageName,
      })
      .then((r) => r.data),

  /** List available credit packages for purchase. */
  getCreditPackages: () =>
    api
      .get<CreditPackageItem[]>('/api/wallet/credits')
      .then((r) => r.data),

  /** Create Stripe PaymentIntent for credit purchase. */
  createCheckout: (packageId: number) =>
    api
      .post<CheckoutResponse>('/api/wallet/credits/checkout', {
        package_id: packageId,
      })
      .then((r) => r.data),

  /** Update auto-refill settings. */
  updateAutoRefill: (enabled: boolean, thresholdCents: number, amountCents: number) =>
    api
      .patch<WalletData>('/api/wallet/auto-refill', {
        enabled,
        threshold_cents: thresholdCents,
        amount_cents: amountCents,
      })
      .then((r) => r.data),
};

File: app/schemas/wallet.py
"""
Pydantic schemas for the Digital Wallet & Subscription System (§6.60).
"""
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class WalletResponse(BaseModel):
    """Wallet balance and package info."""
    id: int
    package: str
    package_credits: float
    purchased_credits: float
    total_credits: float
    auto_refill_enabled: bool
    auto_refill_threshold_cents: int
    auto_refill_amount_cents: int

    model_config = ConfigDict(from_attributes=True)


class PackageTierResponse(BaseModel):
    """Available subscription tier."""
    id: int
    name: str
    monthly_credits: float
    price_cents: int
    is_active: bool

    model_config = ConfigDict(from_attributes=True)


class WalletTransactionResponse(BaseModel):
    """Single transaction in the immutable ledger."""
    id: int
    transaction_type: str
    amount: float
    balance_after: float
    reference_id: str | None = None
    payment_method: str | None = None
    note: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WalletTransactionList(BaseModel):
    """Paginated transaction list."""
    items: list[WalletTransactionResponse]
    total: int


class CreditPackageResponse(BaseModel):
    """Purchasable credit bundle."""
    id: int
    name: str
    credits: float
    price_cents: int

    model_config = ConfigDict(from_attributes=True)


class PackageEnrollRequest(BaseModel):
    """Request to change subscription tier."""
    package_name: str = Field(..., description="Tier name: free, standard, or premium")


class CreditPurchaseRequest(BaseModel):
    """Request to buy a credit pack via Stripe."""
    package_id: int = Field(..., gt=0)


class CheckoutResponse(BaseModel):
    """Stripe PaymentIntent client secret for frontend."""
    client_secret: str
    publishable_key: str


class AutoRefillRequest(BaseModel):
    """Configure wallet auto-refill settings."""
    enabled: bool
    threshold_cents: int = Field(0, ge=0)
    amount_cents: int = Field(500, ge=100)
